pkg/logstructured/logstructured.go

```diff
@@ -294,28 +300,57 @@ func (l *LogStructured) ttlEvents(ctx context.Context) chan *server.Event {
 				}
 			}
 		}
+
+		if ctx.Err() == nil {
+			logrus.Error("ttl events watch channel was closed due to slow processing")
+		}
 	}()
 
 	return result
 }
 
 func (l *LogStructured) ttl(ctx context.Context) {
 	// vary naive TTL support
-	mutex := &sync.Mutex{}
-	for event := range l.ttlEvents(ctx) {
-		go func(event *server.Event) {
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(time.Duration(event.KV.Lease) * time.Second):
-			}
-			mutex.Lock()
-			if _, _, _, err := l.Delete(ctx, event.KV.Key, event.KV.ModRevision); err != nil {
-				logrus.Errorf("failed to delete expired key: %v", err)
-			}
-			mutex.Unlock()
-		}(event)
+	queue := workqueue.NewDelayingQueue()
+	go func() {
+		for l.handleTTLEvents(ctx, queue) {
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			queue.ShutDown()
+			return
+		default:
+		}
+
+		for event := range l.ttlEvents(ctx) {
+			queue.AddAfter(event, time.Duration(event.KV.Lease)*time.Second)
+		}
+	}
+}
+
+func (l *LogStructured) handleTTLEvents(ctx context.Context, queue workqueue.DelayingInterface) bool {
+	item, shutdown := queue.Get()
+	if shutdown {
+		logrus.Info("ttl events work queue has shut down")
+		return false
+	}
+
+	defer queue.Done(item)
+	event, ok := item.(*server.Event)
+	if !ok {
+		return true
+	}
+
+	logrus.Tracef("handle ttl events key: %v, modRev: %v", event.KV.Key, event.KV.ModRevision)
+
+	if _, _, _, err := l.Delete(ctx, event.KV.Key, event.KV.ModRevision); err != nil {
```

Comment:
You should manually test to see what happens if the key is deleted or updated before the TTL expires, I think the current state of retrying on all errors is incorrect. I think you need to handle `ErrKeyExists` and maybe also `ErrCompacted` here by not retrying. It is possible that the key was already deleted, or it was updated to no longer have a TTL.

Reply:
Yeah, I have already tested for different cases. Also, let's explore the logic of of Delete from a code perspective，I manually added comments in the following code.

```
func (l *LogStructured) Delete(ctx context.Context, key string, revision int64) (revRet int64, kvRet *server.KeyValue, deletedRet bool, errRet error) {
	defer func() {
		l.adjustRevision(ctx, &revRet)
		logrus.Tracef("DELETE %s, rev=%d => rev=%d, kv=%v, deleted=%v, err=%v", key, revision, revRet, kvRet != nil, deletedRet, errRet)
	}()
        
        // Case 1: Retry
        // Obtain the specified key from the backend, and if it fails, return error. 
        // This error may contain network or database errors.
	rev, event, err := l.get(ctx, key, "", 1, 0, true)
	if err != nil {
		return 0, nil, false, err
	}
         
        // Case 2: Don't Retry
        // If the key don't exist in DB,  event will be nil. So the returned error is nil.
	if event == nil {
		return rev, nil, true, nil
	}
        
        // Case 3: Don't Retry
        // If the event is already deleted, the deleted column in DB will be set to 1.
        // When trying to delete again, the returned error is nil. 
	if event.Delete {
		return rev, event.KV, true, nil
	}
        
        // Case 4: Don't Retry
        // If the event is updated, the updated event modRevision will be not equal to the previous revision.
        // So the returned error is nil.
	if revision != 0 && event.KV.ModRevision != revision {
		return rev, event.KV, false, nil
	}

	deleteEvent := &server.Event{
		Delete: true,
		KV:     event.KV,
		PrevKV: event.KV,
	}

	rev, err = l.log.Append(ctx, deleteEvent)
	if err != nil {
		// If error on Append we assume it's a UNIQUE constraint error, so we fetch the latest (if we can)
		// and return that the delete failed
                 
                // Case 5: Don't Retry (Here, I have some questions.)
		latestRev, latestEvent, latestErr := l.get(ctx, key, "", 1, 0, true)
                 // Q1: If latestErr != nil or latestEvent == nil, Why return deleted == false and error == nil ?
                 // For this case:
                 //       If latestErr != nil, do we return latestErr ?
                 //       if llatestErr == nil and latestEvent == nil, do we already return error with err?
		if latestErr != nil || latestEvent == nil {
			return rev, event.KV, false, nil
		}
                // Q2:  If the event was already deleted, why also return deleted == false and error == nil ?
                // For this case:
                //         do we need return deleted == true and error == nil ?
		return latestRev, latestEvent.KV, false, nil
	}
        // Case 6:  Don't Retry.
        // Maybe it will only be returned when error == nil
	return rev, event.KV, true, err
}

```

Reply:
right, it's been a while since I did a deep dive into that bit of the code. You are correct that it appears that the only thing that should return `errRet != nil` would be an error on the initial get. All other paths swallow the error and just return `deletedRet = false` to indicate that the key was not deleted. So I guess we just want to requeue on any error.